Add plugin: Enhanced Canvas

RobertttBS opened the pull request:
# I am submitting a new Community Plugin

## Repo URL

<!--- Paste a link to your repo here for easy access -->
Link to my plugin: https://github.com/RobertttBS/obsidian-enhanced-canvas.git

## Release Checklist
- [x] I have tested the plugin on
  - [ ]  Windows
  - [x]  macOS
  - [x]  Linux
  - [ ]  Android _(if applicable)_
  - [x]  iOS _(if applicable)_
- [x] My GitHub release contains all required files (as individual files, not just in the source.zip / source.tar.gz)
  - [x] `main.js`
  - [x] `manifest.json`
  - [x] `styles.css` _(optional)_
- [x] GitHub release name matches the exact version number specified in my manifest.json (_**Note:** Use the exact version number, don't include a prefix `v`_)
- [x] The `id` in my `manifest.json` matches the `id` in the `community-plugins.json` file.
- [x] My README.md describes the plugin's purpose and provides clear usage instructions.
- [x] I have read the developer policies at https://docs.obsidian.md/Developer+policies, and have assessed my plugins's adherence to these policies.
- [x] I have read the tips in https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines and have self-reviewed my plugin to avoid these common pitfalls.
- [x] I have added a license in the LICENSE file.
- [x] My project respects and is compatible with the original license of any code from other plugins that I'm using.
      I have given proper attribution to these other projects in my `README.md`.

Zachatoo commented:
[# Obsidian - Enhanced Canvas](https://github.com/RobertttBS/obsidian-enhanced-canvas/blob/59c5e2bc709895f70856d6979db7e66a359ab1f3/README.md#L1)
Naming something "Obsidian xyz" is reserved for first party products we create, please change this header.

Remove the `styles.css` file from your release, since it's empty it's just unnecessary bloat that Obsidian would download.

["minAppVersion": "0.15.0",](https://github.com/RobertttBS/obsidian-enhanced-canvas/blob/59c5e2bc709895f70856d6979db7e66a359ab1f3/manifest.json#L5)
You are using newer API's, set this to the latest public build number.

[console.log('Loading Enhanced Canvas');](https://github.com/RobertttBS/obsidian-enhanced-canvas/blob/59c5e2bc709895f70856d6979db7e66a359ab1f3/main.ts#L219), [console.log('patched edge');](https://github.com/RobertttBS/obsidian-enhanced-canvas/blob/59c5e2bc709895f70856d6979db7e66a359ab1f3/main.ts#L446), [console.log('patched canvas');](https://github.com/RobertttBS/obsidian-enhanced-canvas/blob/59c5e2bc709895f70856d6979db7e66a359ab1f3/main.ts#L512)
Please avoid unnecessary logging. If you need it for debugging purposes, add a check to only log during development.

[around(this.app.fileManager.constructor.prototype, {](https://github.com/RobertttBS/obsidian-enhanced-canvas/blob/59c5e2bc709895f70856d6979db7e66a359ab1f3/main.ts#L280), [around(this.app.fileManager.constructor.prototype, {](https://github.com/RobertttBS/obsidian-enhanced-canvas/blob/59c5e2bc709895f70856d6979db7e66a359ab1f3/main.ts#L290), [around(edge.constructor.prototype, {](https://github.com/RobertttBS/obsidian-enhanced-canvas/blob/59c5e2bc709895f70856d6979db7e66a359ab1f3/main.ts#L436), [around(canvas.constructor.prototype, {](https://github.com/RobertttBS/obsidian-enhanced-canvas/blob/59c5e2bc709895f70856d6979db7e66a359ab1f3/main.ts#L465)
Monkey-patching need to be registered, so they can be properly unloaded when the plugin is disabled. Please see documentation around [events](https://docs.obsidian.md/Plugins/Events) and the monkey-patch documentation.

[this.app.workspace.onLayoutReady(() => {](https://github.com/RobertttBS/obsidian-enhanced-canvas/blob/59c5e2bc709895f70856d6979db7e66a359ab1f3/main.ts#L515-L524)
What is the intention of this? `patchChanges()` always returns a falsy value (false or undefined) so your code can be simplified to the following, which I don't think is what you are intending. If you only want to patch the `canvas` prototype once, then you'll need a different approach.
```ts
this.app.workspace.onLayoutReady(() => {
	patchCanvas();
	this.registerEvent(this.app.workspace.on("layout-change", () => {
		patchCanvas();
	}));
});
```

RobertttBS commented:
Thank you for your detailed review.

I have implemented all the suggested modifications:
- [x] Changed the README.md header from # Obsidian - Enhanced Canvas to #
Enhanced Canvas
- [x] Removed the styles.css file from the release package
- [x] Updated the "minAppVersion" in manifest.json from 0.15.0 to 1.7.7
- [x] Ensured proper unregistration of all monkey-patching modifications
when the plugin is disabled
- [x] Refactored the canvas prototype patching logic using the
active-leaf-change event handler. It listens for 'active-leaf-change'
events to patch the canvas. Once the patch is complete, the listener is
removed:

Regarding the last point, I've implemented the following approach:
```
const layoutChangeHandler = () => {
    if (patchCanvas()) {
        // Remove the event listener once canvas is successfully patched
        plugin.app.workspace.off('active-leaf-change', layoutChangeHandler);
    }
};

plugin.app.workspace.on('active-leaf-change', layoutChangeHandler);
```

Zach Young ***@***.***> 於 2025年1月25日 週六 上午10:56寫道：

> # Obsidian - Enhanced Canvas
> <https://github.com/RobertttBS/obsidian-enhanced-canvas/blob/59c5e2bc709895f70856d6979db7e66a359ab1f3/README.md#L1>
> Naming something "Obsidian xyz" is reserved for first party products we
> create, please change this header.
>
> Remove the styles.css file from your release, since it's empty it's just
> unnecessary bloat that Obsidian would download.
>
> "minAppVersion": "0.15.0",
> <https://github.com/RobertttBS/obsidian-enhanced-canvas/blob/59c5e2bc709895f70856d6979db7e66a359ab1f3/manifest.json#L5>
> You are using newer API's, set this to the latest public build number.
>
> console.log('Loading Enhanced Canvas');
> <https://github.com/RobertttBS/obsidian-enhanced-canvas/blob/59c5e2bc709895f70856d6979db7e66a359ab1f3/main.ts#L219>,
> console.log('patched edge');
> <https://github.com/RobertttBS/obsidian-enhanced-canvas/blob/59c5e2bc709895f70856d6979db7e66a359ab1f3/main.ts#L446>,
> console.log('patched canvas');
> <https://github.com/RobertttBS/obsidian-enhanced-canvas/blob/59c5e2bc709895f70856d6979db7e66a359ab1f3/main.ts#L512>
> Please avoid unnecessary logging. If you need it for debugging purposes,
> add a check to only log during development.
>
> around(this.app.fileManager.constructor.prototype, {
> <https://github.com/RobertttBS/obsidian-enhanced-canvas/blob/59c5e2bc709895f70856d6979db7e66a359ab1f3/main.ts#L280>,
> around(this.app.fileManager.constructor.prototype, {
> <https://github.com/RobertttBS/obsidian-enhanced-canvas/blob/59c5e2bc709895f70856d6979db7e66a359ab1f3/main.ts#L290>,
> around(edge.constructor.prototype, {
> <https://github.com/RobertttBS/obsidian-enhanced-canvas/blob/59c5e2bc709895f70856d6979db7e66a359ab1f3/main.ts#L436>,
> around(canvas.constructor.prototype, {
> <https://github.com/RobertttBS/obsidian-enhanced-canvas/blob/59c5e2bc709895f70856d6979db7e66a359ab1f3/main.ts#L465>
> Monkey-patching need to be registered, so they can be properly unloaded
> when the plugin is disabled. Please see documentation around events
> <https://docs.obsidian.md/Plugins/Events> and the monkey-patch
> documentation.
>
> this.app.workspace.onLayoutReady(() => {
> <https://github.com/RobertttBS/obsidian-enhanced-canvas/blob/59c5e2bc709895f70856d6979db7e66a359ab1f3/main.ts#L515-L524>
> What is the intention of this? patchChanges() always returns a falsy
> value (false or undefined) so your code can be simplified to the following,
> which I don't think is what you are intending. If you only want to patch
> the canvas prototype once, then you'll need a different approach.
>
> this.app.workspace.onLayoutReady(() => {
> 	patchCanvas();
> 	this.registerEvent(this.app.workspace.on("layout-change", () => {
> 		patchCanvas();
> 	}));});
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/obsidianmd/obsidian-releases/pull/5115#issuecomment-2613742979>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMICFGWQZN76BJNEXTOXYFD2ML4PDAVCNFSM6AAAAABVAIJ4RCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMMJTG42DEOJXHE>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>

joethei commented:
[canvas](https://github.com/RobertttBS/obsidian-enhanced-canvas/blob/36d110869b2c36b0903cdb977d90c3572cd388d9/main.ts#L307)
Canvas is a product name and is capitalized.

RobertttBS commented:
Thanks for your review, I've done the modification. I've capitalized
"Canvas" as you pointed out.

Johannes Theiner ***@***.***> 於 2025年1月27日 週一 下午7:03寫道：

> canvas
> <https://github.com/RobertttBS/obsidian-enhanced-canvas/blob/36d110869b2c36b0903cdb977d90c3572cd388d9/main.ts#L307>
> Canvas is a product name and is capitalized.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/obsidianmd/obsidian-releases/pull/5115#issuecomment-2615448514>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMICFGUF2L25IDWM447CKDT2MYHAFAVCNFSM6AAAAABVAIJ4RCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMMJVGQ2DQNJRGQ>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>

RobertttBS commented:
Fixed. Thanks.

謝柏陞 ***@***.***> 於 2025年1月28日 週二 上午1:09寫道：

> Thanks for your review, I've done the modification. I've capitalized
> "Canvas" as you pointed out.
>
> Johannes Theiner ***@***.***> 於 2025年1月27日 週一 下午7:03寫道：
>
>> canvas
>> <https://github.com/RobertttBS/obsidian-enhanced-canvas/blob/36d110869b2c36b0903cdb977d90c3572cd388d9/main.ts#L307>
>> Canvas is a product name and is capitalized.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/obsidianmd/obsidian-releases/pull/5115#issuecomment-2615448514>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AMICFGUF2L25IDWM447CKDT2MYHAFAVCNFSM6AAAAABVAIJ4RCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMMJVGQ2DQNJRGQ>
>> .
>> You are receiving this because you modified the open/close state.Message
>> ID: ***@***.***>
>>
>
